docs/dev/datastream_execution_mode.md

```diff
@@ -0,0 +1,377 @@
+nav-pos: 1
+---
+<!--
+Licensed to the Apache Software Foundation (ASF) under one
+or more contributor license agreements.  See the NOTICE file
+distributed with this work for additional information
+regarding copyright ownership.  The ASF licenses this file
+to you under the Apache License, Version 2.0 (the
+"License"); you may not use this file except in compliance
+with the License.  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing,
+software distributed under the License is distributed on an
+"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+KIND, either express or implied.  See the License for the
+specific language governing permissions and limitations
+under the License.
+-->
+
+The DataStream API supports different runtime execution modes from which you
+can choose depending on the requirements of your use case and the
+characteristics of your job.
+
+There is the "classic" execution behavior of the DataStream API, which we call
+`STREAMING` execution mode. This should be used for unbounded jobs that require
+continuous incremental processing and are expected to stay online indefinitely.
+
+Additionally, there is a batch-style execution mode that we call `BATCH`
+execution mode. This executes jobs in a way that is more reminiscent of batch
+processing frameworks such as MapReduce. This should be used for bounded jobs
+for which you have a known fixed input and which do not run continuously.
+
+Apache Flink's unified approach to stream and batch processing means that a
+DataStream application executed over bounded input will produce the same
+results regardless of the configured execution mode.  By enabling `BATCH`
+execution, we allow Flink to apply additional optimizations that we can only do
+work when we know that our input is bounded. For example, different
+join/aggregation strategies can be used, in addition to a different shuffle
+implementation that allows more efficient failure recovery behavior. We will go
+into some of the details of the execution behavior below.
+
+* This will be replaced by the TOC
+{:toc}
+
+## When can/should I use BATCH execution mode?
+
+The `BATCH` execution mode can only be used for Jobs/Flink Programs that are
+_bounded_. Boundedness is a property of a data source that tells us whether all
+the input coming from that source is known before execution or whether new data
+will show up, potentially indefinitely. A job, in turn, is bounded if all its
+sources are bounded, and unbounded otherwise.
+
+`STREAMING` execution mode, on the other hand, can be used for both bounded and
+unbounded jobs.
+
+As a rule of thumb, you should be using `BATCH` execution mode when your program
+is bounded because this will be more efficient. You have to use `STREAMING`
+execution mode when your program is unbounded because only this mode is general
+enough to be able to deal with continuous data streams.
```

Comment:
Here it sounds as if it does not really matter whether to choose `BATCH` or `STREAMING` for a bounded job from a correctness perspective. However, the `FileSink` won't commit the in-progress files at the end of the program when using the `STREAMING` execution mode. It might be worthwhile to document this behaviour somewhere.

Reply:
Yes, this is unfortunate. Though the fact that we cannot do checkpoints as soon as at least one task has finished, which in turn means that we can't get a "final" checkpoint has been a feature/bug of DataStream execution since the beginning. I wouldn't document it here but we can think about adding this to a general "caveats" section. I'm sure there would be other corner cases that are worth documenting 😅